✨Feat: 이미지 업로드 기능 추가

lazzzykim opened the pull request:
## 어떤 의도로 코드를 작성했는지 적어주세요
* 게시물 작성 시 이미지를 업로드 하는 기능을 추가했습니다.
* presigned url 방식을 적용해 서버에 이미지를 저장하지 않고 클라이언트가 S3 버킷에 바로 접근하여 업로드 할 수 있습니다.
* 현재 presigned url의 만료기간이 현재 5분으로 설정되어 있어 5분후에 url이 만료됩니다.
* presigned url을 사용하면 S3에서 버킷에 권한 설정대로 허가된 유저만 사용할 수 있기때문에 서버단에서 인증하는 과정을 생략할 수 있습니다.
* 서버를 거치지 않고 업로드가 되기때문에 서버의 리소스를 사용하지 않고, 이중작업을 방지할 수 있어 (서버에 업로드 후 버킷에 또 업로드 하는) 성능개선의 효과가 있을것으로 예상됩니다.

## 주요 변경점
어떤 부분이 변경되었는지 적어주세요
* Image 도메인이 추가되었습니다. controller와 service에 생성, 조회하는 로직을 추가했습니다.

## ETC
리뷰어들에게 어떤 점을 검토해야하는지 알려주세요
* 현재 프론트에서 이미지 조회 기능은 추가되지 않았지만 백에서 추가적으로 수정할 내용은 없을 것으로 판단해 PR올립니다.
* 추가적으로 진행되는 이미지 관련 기능은 프론트 repository에서 진행하겠습니다

lazzzykim committed:
✨Feat: Coupon 기능 추가

AppliedUserRepository: 중복응모를 막기 위해 add 함수 추가
Coupon: coupon Entity작성
CouponController: 쿠폰발급 엔드포인트 생성
CouponRepository: Coupon Repository 작성
CouponServiceImpl: 쿠폰 발급요청시 유저, 중복 응모, count 검증 후 쿠폰 생성 및 발급 로직 작성
CouponUtility: 쿠폰번호를 16자리로 생성하기 위한 메서드 작성
RedisCouponRepository: 생성시 동시성 문제를 해결하기 위해 redis incr 기능 추가
CreateCouponRequest: 쿠폰 생성시 필요한 Request 작성

resolves: #60

lazzzykim committed:
✨Feat: Coupon 만료 기한 추가

Coupon: 만료기한을 저장하는 couponExp 추가
CouponServiceImpl: request의 couponExp만큼 생성 시 만료기한이 설정되는 기능 추가
CouponExpiration: 자정에 만료된 쿠폰이 삭제되는 로직 작성
CouponRepository: 쿠폰의 만료일자를 조회하는 함수 작성

Ref: #12

lazzzykim committed:
Merge remote-tracking branch 'origin/dev' into feature/s3

lazzzykim committed:
✨Feat: AWS S3 이미지업로드 기능추가

AwsS3Config: Config 파일 작성
AwsS3Service: 이미지 업로드 메서드 작성

resolves: #9

lazzzykim committed:
🔨Refactor: AWS S3 이미지업로드 기능추가

AwsS3Config: Config 파일 수정
AwsS3Service: 이미지 업로드 presigned url 적용한 코드 작성

ref: #9

lazzzykim committed:
Merge remote-tracking branch 'origin/dev' into feature/s3

lazzzykim committed:
🔨Refactor: AWS S3 이미지업로드 기능추가

AwsS3Config: Config 파일 수정
AwsS3Service: 이미지 업로드 presigned url 적용한 코드 작성

ref: #9

lazzzykim committed:
🔨Refactor: AWS S3 이미지업로드 기능

AwsS3Config: aws 연결 설정을 위해 config 작성
AwsS3Service: 이미지 업로드와 조회하는 로직 작성
build.gradle.kts: aws 관련 의존성 추가
ImageController: 이미지 등록, 조회 엔드포인트 추가
Post: imageUrl 컬럼을 추가해 이미지URL이 저장됨
PostListResponse: imageUrl 추가
PostRepositoryImpl: getPostsContents에 imageUrl 추가
PostResponse: imageUrl 추가
PostServiceImpl: createPost 메서드에 request.imageUrl 추가
SecurityConfig: allowedUrls에 Image 추가

ref: #9

lazzzykim committed:
Merge remote-tracking branch 'origin/dev' into feature/s3

# Conflicts:
#	src/main/kotlin/com/teamsparta/buysell/domain/post/dto/response/PostListResponse.kt
#	src/main/kotlin/com/teamsparta/buysell/domain/post/dto/response/PostResponse.kt
#	src/main/kotlin/com/teamsparta/buysell/domain/post/model/Post.kt
#	src/main/kotlin/com/teamsparta/buysell/domain/post/repository/PostRepositoryImpl.kt
#	src/main/kotlin/com/teamsparta/buysell/infra/security/SecurityConfig.kt

lazzzykim committed:
🔨Refactor: AWS S3 이미지업로드 기능

AwsS3Config: aws 연결 설정을 위해 config 작성
AwsS3Service: 이미지 업로드와 조회하는 로직 작성
build.gradle.kts: aws 관련 의존성 추가
ImageController: 이미지 등록, 조회 엔드포인트 추가
Post: imageUrl 컬럼을 추가해 이미지URL이 저장됨
PostListResponse: imageUrl 추가
PostRepositoryImpl: getPostsContents에 imageUrl 추가
PostResponse: imageUrl 추가
PostServiceImpl: createPost 메서드에 request.imageUrl 추가
SecurityConfig: allowedUrls에 Image 추가

ref: #9

lazzzykim committed:
🔨Refactor: AWS S3 이미지업로드 기능

Comment, Post Test code에 imgurl 추가

ref: #9

lazzzykim committed:
🔨Refactor: AWS S3 이미지업로드 기능

CreateCouponTest: 전체 주석처리 (현재 RDS에 redis적용 안되있어서 테스트가 실패함)

ref: #9

HwangSeungHyeon commented on src/main/kotlin/com/teamsparta/buysell/domain/coupon/service/CouponServiceImpl.kt:
```diff
@@ -0,0 +1,55 @@
+    @Transactional
+    override fun createCoupon(request: CreateCouponRequest, memberId: Int): MessageResponse {
+        val member = memberRepository.findByIdOrNull(memberId)
+            ?: throw IllegalArgumentException("Invalid member")
```
혹시 ModelNotFoundException 대신 쓰신 이유가 있으신가요??

lazzzykim replied:
유저 중복 검증이라고 계속 생각하다보니 Illegar을 사용했는데 생각해보니 유저를 조회하고 없으면 ModelNotFound가 맞는거 같습니다..ㅋㅋ
수정했습니다 확인부탁드립니다~

HwangSeungHyeon commented on src/main/kotlin/com/teamsparta/buysell/domain/image/controller/ImageController.kt:
```diff
@@ -0,0 +1,38 @@
+import org.springframework.web.bind.annotation.*
+import java.io.IOException
+
+@CrossOrigin(origins = ["http://localhost:3000"])
```
localhost를 사용해도 배포 서버에서 사용할 수 있나요?

lazzzykim replied:
cors 에러 났을 때 이것저것 시도해본 잔해가.. 수정했습니다.

lazzzykim committed:
🔨Refactor: AWS S3 이미지업로드 기능

CouponServiceImpl: IllegarArgumentsException -> ModelNotFoundException으로 수정
ImageController: CrossOrigin 어노테이션 삭제
AwsS3Service.kt: 사용하지 않는 클래스 삭제

ref: #9

HwangSeungHyeon reviewed (approved):
고생 많으셨습니다.
